docs/design.md

```diff
@@ -0,0 +1,194 @@
+specification of how to update the repository. For each of these, it:
+
+ - calculates which image repositories need to be scanned, and the
+   policies for updating them, according to the specification given in
+   the `ImageAutomation` and git repository;
+ - consults the policies it created to determine updates to perform;
+ - creates and manages `UpdateJob` resources to run the updates.
+
+### Specification
+
+To be designed -- see the notes below.
+
+### Design notes
+
+There's a large design space for the automation, along various axes:
+
+ - where does the specification for what is automated live -- in git
+   or in a resource?
+ - is the specification part of the resource/file it automates, or
+   separate?
+ - does the specification name all the things to which it applies; or
+   does it work with rules or patterns?
+
+So, for example, one design could lie at this point in the space: an
+`ImageAutomation` resource names a git repository, a workload object,
+and an `ImagePolicy`; every time the policy selects an image that does
+not match what is given for the workload resource in git, a job is
+created to update it.
+
+It is easy to see how to implement this design, since everything is
+totally explicit -- you just do what the resources say. But it's not
+great for the user, because they have to spend time spelling it all
+out for the controller, and do the work of keeping the automation and
+policy objects in the cluster up to date with what's in the git repo
+(one way to do that is to keep them in the git repo and let them be
+synced; but in general, the workloads are not going to run in the same
+place as the automation, so it would be fiddly to keep these in the
+same place).
+
+A design in the other direction would be to expect annotations,
+similar to those used by Flux v1, on workloads to be automated. The
+controller would interpret those to determine which image repositories
+and policies are needed.
+
+This might be tricky when the automation is managed by a different
+team to that in charge of the application configuration -- in that
+scenario, the annotations would have to be carefully applied after the
+fact (perhaps with a kustomization), which may as well mean the
+annotations are kept in a separate file.
+
+## Easy (Flux v1) mode
+
+This design includes an "easy mode" for simple deployments, which
+mimics closely the behaviour of the automation in Flux v1. Creating an
+`ImageAutomation` resource that analyses the annotations in git
+repository will give much the same results as in Flux v1.
+
+## Extension points
+
+**Actions other than UpdateJob**
+
+As a way in to other systems, e.g., Tekton, the `ImageAutomation` job
+could specify an action other than creating an `UpdateJob` (for
+Tekton, this might be creating a `TaskRun`). The impedence here is how
+to package the information about what to update -- this will need to
+be thought through for each such integration.
+
+**Supply a different image to UpdateJob**
```

Comment:
Would it be inside or outside of the scope to let `UpdateJob` patch manifests in git for other fields than image-related ones?

Reply:
The rationale for `UpdateJob` is exactly to run arbitrary modifications to a git repository.